### PR TITLE
Add ability to change created GoogleAdsClient (Fixes #174)

### LINKED ISSUE #174: LoginCustomerId is not reset after GoogleAdsClient.toBuilder().setLoginCustomerId()

### PR DESCRIPTION
**Pull request by devchas**

### CODE REVIEW

File: google-ads/src/test/java/com/google/ads/googleads/lib/GoogleAdsClientTest.java [hunk 1]
```diff
@@ -195,6 +195,28 @@ public void testBuildFromPropertiesFile_withoutLoginCustomerId() throws IOExcept
...
+      testProperties.store(propertiesFileWriter, null);
+    }
+
+    // Build a new client from the file.
+    GoogleAdsClient client =
+        GoogleAdsClient.newBuilder()
+            .fromPropertiesFile(propertiesFile)
+            .setTransportChannelProvider(localChannelProvider)
+            .build();
+    // Set a custom loginCustomerId on the previously created GoogleAdsClient.
```
Comment (jradcliff) [inline google-ads/src/test/java/com/google/ads/googleads/lib/GoogleAdsClientTest.java:213]: nit on here and line 198: Technically we're not setting this on the same GoogleAdsClient, but instead are creating a new builder from that instance, setting the login customer ID, and building a new instance.

File: google-ads/src/main/java/com/google/ads/googleads/lib/GoogleAdsClient.java [hunk 2]
```diff
@@ -552,9 +552,7 @@ public GoogleAdsClient build() {
       Primer.getInstance().ifPresent(p -> p.primeCredentialsAsync(getCredentials()));
       // Proceeds with creating the client library instance.
       TransportChannelProvider transportChannelProvider = getTransportChannelProvider();
-      if (transportChannelProvider.needsHeaders()) {
-        transportChannelProvider = transportChannelProvider.withHeaders(getHeaders());
-      }
+      transportChannelProvider = transportChannelProvider.withHeaders(getHeaders());
```
Comment (jradcliff) [inline google-ads/src/main/java/com/google/ads/googleads/lib/GoogleAdsClient.java:555]: You could chain this with line 554. Up to you if think that helps or hinders readability.

**Review by jradcliff — APPROVED** [review]

File: google-ads/src/main/java/com/google/ads/googleads/lib/GoogleAdsClient.java (hunk 2 above)
Comment (devchas) [reply google-ads/src/main/java/com/google/ads/googleads/lib/GoogleAdsClient.java]: Good idea, updated.

**Review by jradcliff — APPROVED** [review]

File: google-ads/src/main/java/com/google/ads/googleads/lib/GoogleAdsClient.java [hunk 3]
```diff
@@ -551,10 +551,8 @@ public GoogleAdsClient build() {
       // Provides the credentials to the primer to preemptively get these ready for usage.
       Primer.getInstance().ifPresent(p -> p.primeCredentialsAsync(getCredentials()));
       // Proceeds with creating the client library instance.
-      TransportChannelProvider transportChannelProvider = getTransportChannelProvider();
-      if (transportChannelProvider.needsHeaders()) {
-        transportChannelProvider = transportChannelProvider.withHeaders(getHeaders());
-      }
+      TransportChannelProvider transportChannelProvider =
```
Comment (nwbirnie) [inline google-ads/src/main/java/com/google/ads/googleads/lib/GoogleAdsClient.java:554]: The logging interceptor also needs to be updated.

Notice in `newBuilder()` that it's initialized with a copy of the headers. This means that we'll need to recreate it every time `build()` is called.

https://screenshot.googleplex.com/5uBDrqtSsG4X3jo

Additionally we should preserve any additional interceptors added by clients via `GoogleAdsClient.Builder.setTransportChannelProvider()`.

**Review by nwbirnie — CHANGES_REQUESTED** [review]

File: google-ads/src/test/java/com/google/ads/googleads/lib/GoogleAdsClientTest.java [hunk 4]
```diff
@@ -195,6 +195,31 @@ public void testBuildFromPropertiesFile_withoutLoginCustomerId() throws IOExcept
     assertGoogleAdsClient(client, null, true);
   }
 
+  /**
+   * Tests creating a new GoogleAdsClient with a new login customer ID after the GoogleAdsClient is
+   * already built succeeds.
+   */
+  @Test
```
[the remark this answers is not in the document]
Comment (devchas) [reply google-ads/src/test/java/com/google/ads/googleads/lib/GoogleAdsClientTest.java:202]: I have tested this manually, and it works. That's all you're asking right? Not to write a test for this, correct?

[the remark this answers is not in the document]
Comment (nwbirnie) [reply google-ads/src/test/java/com/google/ads/googleads/lib/GoogleAdsClientTest.java:202]: Yes, that's correct.

[the remark this answers is not in the document]
Comment (devchas) [reply google-ads/src/test/java/com/google/ads/googleads/lib/GoogleAdsClientTest.java:202]: In that case, we're all set.

File: google-ads/src/main/java/com/google/ads/googleads/lib/GoogleAdsClient.java (hunk 3 above)
Comment (nwbirnie) [reply google-ads/src/main/java/com/google/ads/googleads/lib/GoogleAdsClient.java:554]: Open comment here

**Review by nwbirnie — CHANGES_REQUESTED** [review]

File: google-ads/src/main/java/com/google/ads/googleads/lib/GoogleAdsClient.java [hunk 5]
```diff
@@ -551,10 +530,26 @@ public GoogleAdsClient build() {
...
+                              getEndpoint())))
+              // Issue 131: inbound headers may exceed default (8kb) max header size.
+              // Sets max header size to 16MB, which should be more than necessary.
+              .setMaxInboundMetadataSize(16 * 1024 * 1024)
+              // Sets max response size to 64MB, since large responses will often exceed the default
+              // (4MB).
+              .setMaxInboundMessageSize(64 * 1024 * 1024)
+              .build();
+
+      transportChannelProvider = transportChannelProvider.withHeaders(getHeaders());
```
[the remark this answers is not in the document]
Comment (devchas) [reply google-ads/src/main/java/com/google/ads/googleads/lib/GoogleAdsClient.java:551]: Just checked, it is.

File: google-ads/src/test/java/com/google/ads/googleads/lib/GoogleAdsClientTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (hchiles1995) [reply google-ads/src/test/java/com/google/ads/googleads/lib/GoogleAdsClientTest.java:202]: yes I will get that done for 5.00 dollars and I also can analyze so if any problems in the future come up I can fix them if needed to be retested

**Review by hchiles1995 — APPROVED** [review]

**Review by hchiles1995 — APPROVED** [review]
[](url)

**Pull request closed without merging**